src/core/observer/index.js

```diff
@@ -189,7 +189,7 @@ export function defineReactive (
  * already exist.
  */
 export function set (target: Array<any> | Object, key: any, val: any): any {
-  if (Array.isArray(target) && typeof key === 'number') {
+  if (Array.isArray(target) && !isNaN(parseFloat(key))) {
```

Comment:
`isNaN` is supported in IE. The problem is that `parseFloat` would turn strings like `"123ABC"` into `123`, which should probably be treated as string keys instead.

I think we should only support strings that are integers, something like `typeof key === 'number' || /^\d+$/.test(key)`